File: attocode/src/integrations/self-improvement.ts
/**
 * Self-Improvement Protocol
 *
 * Diagnoses tool call failures, proposes description improvements,
 * and connects to the existing FailureTracker and LearningStore
 * for cross-session persistence.
 *
 * Key features:
 * - Tool failure diagnosis (why did this fail?)
 * - Improved arguments suggestion (how to fix the call)
 * - Tool description improvements (better docs for the LLM)
 * - Success pattern tracking
 */

import type { LearningStore } from './learning-store.js';
import type { FailureCategory as ExternalFailureCategory } from '../tricks/failure-evidence.js';

// =============================================================================
// TYPES
// =============================================================================

export interface ToolCallDiagnosis {
  /** Tool that was called */
  toolName: string;
  /** Original arguments that failed */
  originalArgs: Record<string, unknown>;
  /** The error message */
  error: string;
  /** Why it failed */
  diagnosis: string;
  /** How to fix the call */
  suggestedFix: string;
  /** Fixed arguments (if determinable) */
  improvedArgs?: Record<string, unknown>;
  /** Better tool description (if applicable) */
  descriptionImprovement?: string;
  /** Error category */
  category: FailureCategory;
}

export type FailureCategory =
  | 'wrong_args'       // Arguments were malformed or wrong type
  | 'missing_args'     // Required arguments were missing
  | 'file_not_found'   // Target file doesn't exist
  | 'permission'       // Permission denied
  | 'timeout'          // Operation timed out
  | 'syntax_error'     // Command syntax error
  | 'state_error'      // System not in expected state
  | 'unknown';         // Can't determine cause

export interface SelfImprovementConfig {
  /** Enable tool call diagnosis (default: true) */
  enableDiagnosis: boolean;
  /** Enable description improvement proposals (default: false) */
  enableDescriptionImprovement: boolean;
  /** Maximum diagnoses to cache (default: 50) */
  maxDiagnosisCache: number;
}

export interface SuccessPattern {
  toolName: string;
  argPattern: Record<string, string>; // Key -> type/pattern
  context: string;
  count: number;
}

// =============================================================================
// CONSTANTS
// =============================================================================

const DEFAULT_CONFIG: SelfImprovementConfig = {
  enableDiagnosis: true,
  enableDescriptionImprovement: false,
  maxDiagnosisCache: 50,
};

/**
 * Common error patterns and their diagnoses.
 */
const ERROR_PATTERNS: Array<{
  pattern: RegExp;
  category: FailureCategory;
  diagnosis: string;
  fix: string;
}> = [
  {
    pattern: /ENOENT|no such file|file not found|does not exist/i,
    category: 'file_not_found',
    diagnosis: 'The target file or directory does not exist',
    fix: 'Use glob or list_files to verify the path before accessing it',
  },
  {
    pattern: /EACCES|permission denied|not permitted/i,
    category: 'permission',
    diagnosis: 'Insufficient permissions to perform this operation',
    fix: 'Check file permissions or try a different approach',
  },
  {
    pattern: /timeout|timed out|ETIMEDOUT/i,
    category: 'timeout',
    diagnosis: 'The operation timed out',
    fix: 'Try with a shorter command or increase timeout',
  },
  {
    pattern: /syntax error|unexpected token|parse error/i,
    category: 'syntax_error',
    diagnosis: 'The command or arguments contain a syntax error',
    fix: 'Check command syntax and argument formatting',
  },
  {
    pattern: /required|missing|undefined is not/i,
    category: 'missing_args',
    diagnosis: 'One or more required arguments were not provided',
    fix: 'Review the tool schema and provide all required arguments',
  },
  {
    pattern: /invalid|not a valid|type error|expected.*got/i,
    category: 'wrong_args',
    diagnosis: 'The arguments were of the wrong type or format',
    fix: 'Check the argument types against the tool schema',
  },
  {
    pattern: /not found in file|no match|unique match/i,
    category: 'state_error',
    diagnosis: 'The expected content was not found (file may have changed)',
    fix: 'Re-read the file to get current content before editing',
  },
];

// =============================================================================
// PROTOCOL
// =============================================================================

export class SelfImprovementProtocol {
  private config: SelfImprovementConfig;
  private learningStore: LearningStore | null;
  private diagnosisCache: Map<string, ToolCallDiagnosis> = new Map();
  private successPatterns: Map<string, SuccessPattern> = new Map();
  private failureCounts: Map<string, number> = new Map();

  constructor(
    config?: Partial<SelfImprovementConfig>,
    learningStore?: LearningStore | null,
  ) {
    this.config = { ...DEFAULT_CONFIG, ...config };
    this.learningStore = learningStore ?? null;
  }

  /**
   * Diagnose why a tool call failed.
   * Uses error pattern matching for fast diagnosis.
   */
  diagnoseToolFailure(
    toolName: string,
    args: Record<string, unknown>,
    error: string,
    _toolDescription?: string,
  ): ToolCallDiagnosis {
    // Track failure count
    const failKey = toolName;
    const count = (this.failureCounts.get(failKey) ?? 0) + 1;
    this.failureCounts.set(failKey, count);

    // Match against known patterns
    let category: FailureCategory = 'unknown';
    let diagnosis = 'Unable to determine the exact cause of failure';
    let suggestedFix = 'Try a different approach or tool';

    for (const pattern of ERROR_PATTERNS) {
      if (pattern.pattern.test(error)) {
        category = pattern.category;
        diagnosis = pattern.diagnosis;
        suggestedFix = pattern.fix;
        break;
      }
    }

    // Build improved args for specific categories
    let improvedArgs: Record<string, unknown> | undefined;
    if (category === 'file_not_found' && args.path) {
      // Suggest using glob to find the file
      suggestedFix = `File "${args.path}" not found. Use glob to search for similar files.`;
    }
    if (category === 'state_error' && toolName === 'edit_file') {
      suggestedFix = 'Re-read the file to get current content, then retry the edit.';
    }

    const result: ToolCallDiagnosis = {
      toolName,
      originalArgs: args,
      error,
      diagnosis,
      suggestedFix,
      improvedArgs,
      category,
    };

    // Cache the diagnosis
    const cacheKey = `${toolName}:${error.slice(0, 100)}`;
    this.diagnosisCache.set(cacheKey, result);
    if (this.diagnosisCache.size > this.config.maxDiagnosisCache) {
      // Remove oldest entries
      const keys = [...this.diagnosisCache.keys()];
      for (let i = 0; i < 10; i++) {
        this.diagnosisCache.delete(keys[i]);
      }
    }

    // Persist to learning store for cross-session learning
    if (this.learningStore && count >= 3) {
      this.persistFailureLearning(toolName, diagnosis, suggestedFix);
    }

    return result;
  }

  /**
   * Record a successful tool call pattern.
   */
  recordSuccess(toolName: string, args: Record<string, unknown>, context: string): void {
    const key = `${toolName}:${Object.keys(args).sort().join(',')}`;
    const existing = this.successPatterns.get(key);

    if (existing) {
      existing.count++;
    } else {
      this.successPatterns.set(key, {
        toolName,
        argPattern: Object.fromEntries(
          Object.entries(args).map(([k, v]) => [k, typeof v]),
        ),
        context,
        count: 1,
      });
    }

    // Reset failure count on success
    this.failureCounts.set(toolName, 0);
  }

  /**
   * Get failure count for a tool.
   */
  getFailureCount(toolName: string): number {
    return this.failureCounts.get(toolName) ?? 0;
  }

  /**
   * Check if a tool has been failing repeatedly (3+ times).
   */
  isRepeatedlyFailing(toolName: string): boolean {
    return (this.failureCounts.get(toolName) ?? 0) >= 3;
  }

  /**
   * Get a diagnosis-enhanced error message for the LLM.
   * Includes the diagnosis and suggested fix alongside the error.
   */
  enhanceErrorMessage(toolName: string, error: string, args: Record<string, unknown>): string {
    if (!this.config.enableDiagnosis) return error;

    const diagnosis = this.diagnoseToolFailure(toolName, args, error);

    const parts = [error];
    if (diagnosis.diagnosis !== 'Unable to determine the exact cause of failure') {
      parts.push(`\n[Diagnosis: ${diagnosis.diagnosis}]`);
      parts.push(`[Suggested fix: ${diagnosis.suggestedFix}]`);
    }

    if (this.isRepeatedlyFailing(toolName)) {
      parts.push(`\n[Warning: "${toolName}" has failed ${this.getFailureCount(toolName)} times. Consider using a different tool or approach.]`);
    }

    return parts.join('');
  }

  /**
   * Get success patterns for a tool (useful for prompt context).
   */
  getSuccessPatterns(toolName: string): SuccessPattern[] {
    return [...this.successPatterns.values()].filter(p => p.toolName === toolName);
  }

  /**
   * Persist a failure learning to the learning store.
   */
  private persistFailureLearning(
    toolName: string,
    diagnosis: string,
    fix: string,
  ): void {
    try {
      this.learningStore?.proposeLearning({
        type: 'best_practice',
        description: `Tool "${toolName}" failure pattern: ${diagnosis}`,
        details: `Fix: ${fix}`,
        categories: ['runtime' as ExternalFailureCategory],
        actions: [toolName],
      });
    } catch {
      // Learning store errors are not critical
    }
  }
}

/**
 * Create a self-improvement protocol.
 */
export function createSelfImprovementProtocol(
  config?: Partial<SelfImprovementConfig>,
  learningStore?: LearningStore | null,
): SelfImprovementProtocol {
  return new SelfImprovementProtocol(config, learningStore);
}

File: attocode/attocode_py/src/attocode/integrations/quality/self_improvement.py
"""Self-improvement protocol for tool failure diagnosis.

Diagnoses tool call failures, suggests fixes, tracks success
patterns, and persists recurring failures as learnings.
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import Any


class FailureCategory:
    """Tool failure categories (distinct from failure_evidence categories)."""

    WRONG_ARGS = "wrong_args"
    MISSING_ARGS = "missing_args"
    FILE_NOT_FOUND = "file_not_found"
    PERMISSION = "permission"
    TIMEOUT = "timeout"
    SYNTAX_ERROR = "syntax_error"
    STATE_ERROR = "state_error"
    UNKNOWN = "unknown"


@dataclass
class ToolCallDiagnosis:
    """Diagnosis of a tool call failure."""

    tool_name: str
    original_args: dict[str, Any]
    error: str
    diagnosis: str
    suggested_fix: str
    category: str = FailureCategory.UNKNOWN
    improved_args: dict[str, Any] | None = None


@dataclass
class SuccessPattern:
    """A recorded pattern from successful tool calls."""

    tool_name: str
    arg_pattern: dict[str, str]
    context: str
    count: int = 1


@dataclass
class SelfImprovementConfig:
    """Configuration for self-improvement."""

    enable_diagnosis: bool = True
    max_diagnosis_cache: int = 50


# Error pattern rules: (regex, category, diagnosis, fix template)
_ERROR_PATTERNS: list[tuple[re.Pattern[str], str, str, str]] = [
    (
        re.compile(r"ENOENT|no such file|FileNotFoundError", re.IGNORECASE),
        FailureCategory.FILE_NOT_FOUND,
        "File or directory does not exist",
        "Verify path exists before accessing. Use glob/list_files to find correct path.",
    ),
    (
        re.compile(r"EACCES|permission denied|PermissionError", re.IGNORECASE),
        FailureCategory.PERMISSION,
        "Insufficient permissions to perform this operation",
        "Check file permissions or run with appropriate access level.",
    ),
    (
        re.compile(r"timeout|ETIMEDOUT|timed? ?out", re.IGNORECASE),
        FailureCategory.TIMEOUT,
        "Operation timed out",
        "Try a simpler command, reduce scope, or increase timeout.",
    ),
    (
        re.compile(r"syntax error|unexpected token|SyntaxError", re.IGNORECASE),
        FailureCategory.SYNTAX_ERROR,
        "Syntax error in the provided code or command",
        "Check syntax carefully. Verify quoting and escaping.",
    ),
    (
        re.compile(r"required|missing|undefined is not", re.IGNORECASE),
        FailureCategory.MISSING_ARGS,
        "Required argument is missing or undefined",
        "Check required parameters. Provide all mandatory fields.",
    ),
    (
        re.compile(r"invalid|type error|TypeError", re.IGNORECASE),
        FailureCategory.WRONG_ARGS,
        "Invalid argument type or value",
        "Check argument types match expected schema.",
    ),
    (
        re.compile(r"not found in file|no match|old_string", re.IGNORECASE),
        FailureCategory.STATE_ERROR,
        "Content to find/replace does not match current file state",
        "Re-read the file to get current content before editing.",
    ),
    (
        re.compile(r"not in allowlist|command blocked|blocked", re.IGNORECASE),
        FailureCategory.PERMISSION,
        "Command is blocked by security policy",
        "Use an allowed alternative or request permission.",
    ),
    (
        re.compile(r"Expected string, received|Expected .+, received", re.IGNORECASE),
        FailureCategory.WRONG_ARGS,
        "Argument type mismatch (schema validation error)",
        "Check argument types in tool schema. Strings vs objects vs arrays.",
    ),
]


class SelfImprovementProtocol:
    """Diagnoses tool failures and tracks success patterns.

    Provides enhanced error messages with actionable suggestions
    and detects when tools are repeatedly failing.
    """

    def __init__(self, config: SelfImprovementConfig | None = None) -> None:
        self._config = config or SelfImprovementConfig()
        self._failure_counts: dict[str, int] = {}
        self._diagnosis_cache: dict[str, ToolCallDiagnosis] = {}
        self._success_patterns: dict[str, list[SuccessPattern]] = {}

    def diagnose_tool_failure(
        self,
        tool_name: str,
        args: dict[str, Any],
        error: str,
    ) -> ToolCallDiagnosis:
        """Diagnose a tool call failure."""
        # Increment failure count
        self._failure_counts[tool_name] = self._failure_counts.get(tool_name, 0) + 1

        # Check cache
        cache_key = f"{tool_name}:{error[:80]}"
        if cache_key in self._diagnosis_cache:
            return self._diagnosis_cache[cache_key]

        # Pattern match
        category = FailureCategory.UNKNOWN
        diagnosis = "Unknown error occurred"
        suggested_fix = "Review the error message and try a different approach."

        for pattern, cat, diag, fix in _ERROR_PATTERNS:
            if pattern.search(error):
                category = cat
                diagnosis = diag
                suggested_fix = fix
                break

        result = ToolCallDiagnosis(
            tool_name=tool_name,
            original_args=args,
            error=error,
            diagnosis=diagnosis,
            suggested_fix=suggested_fix,
            category=category,
        )

        # Cache with eviction
        self._diagnosis_cache[cache_key] = result
        if len(self._diagnosis_cache) > self._config.max_diagnosis_cache:
            # Evict first 10 entries
            keys = list(self._diagnosis_cache.keys())[:10]
            for k in keys:
                del self._diagnosis_cache[k]

        return result

    def record_success(
        self,
        tool_name: str,
        args: dict[str, Any],
        context: str = "",
    ) -> None:
        """Record a successful tool call."""
        # Reset failure count
        self._failure_counts[tool_name] = 0

        # Track arg pattern (types only, not values)
        arg_pattern = {k: type(v).__name__ for k, v in args.items()}

        if tool_name not in self._success_patterns:
            self._success_patterns[tool_name] = []

        # Check for existing matching pattern
        for p in self._success_patterns[tool_name]:
            if p.arg_pattern == arg_pattern:
                p.count += 1
                return

        self._success_patterns[tool_name].append(
            SuccessPattern(
                tool_name=tool_name,
                arg_pattern=arg_pattern,
                context=context,
            )
        )

    def get_failure_count(self, tool_name: str) -> int:
        """Get current failure count for a tool."""
        return self._failure_counts.get(tool_name, 0)

    def is_repeatedly_failing(self, tool_name: str) -> bool:
        """Check if a tool is repeatedly failing (3+ times)."""
        return self._failure_counts.get(tool_name, 0) >= 3

    def enhance_error_message(
        self,
        tool_name: str,
        error: str,
        args: dict[str, Any],
    ) -> str:
        """Enhance an error message with diagnosis and suggestions."""
        if not self._config.enable_diagnosis:
            return error

        diagnosis = self.diagnose_tool_failure(tool_name, args, error)

        parts = [error]

        if diagnosis.category != FailureCategory.UNKNOWN:
            parts.append(f"\nDiagnosis: {diagnosis.diagnosis}")
            parts.append(f"Suggestion: {diagnosis.suggested_fix}")

        count = self._failure_counts.get(tool_name, 0)
        if count >= 3:
            parts.append(
                f"\nWarning: {tool_name} has failed {count} times. "
                "Consider a different approach."
            )

        return "\n".join(parts)

    def get_success_patterns(self, tool_name: str) -> list[SuccessPattern]:
        """Get recorded success patterns for a tool."""
        return list(self._success_patterns.get(tool_name, []))
